### PR TITLE
Null sourced queries

### PR DESCRIPTION
**Pull request by cmoesel**
Fix handling of null-sourced queries so they return null as described here:
https://jira.hl7.org/browse/FHIR-40225

Also refactor multi-source queries to be more efficient, so we avoid executing sources multiple times.

And re-enable and improve some skipped tests.

Fixes https://github.com/projecttacoma/fqm-execution/issues/172

Pull requests into cql-execution require the following.
Submitter and reviewer should ✔ when done.
For items that are not-applicable, mark "N/A" and ✔.

[CDS Connect](https://cds.ahrq.gov/cdsconnect) and [Bonnie](https://github.com/projecttacoma/bonnie) are the main users of this repository.
It is strongly recommended to include a person from each of those projects as a reviewer.

**Submitter:**
- [x] This pull request describes why these changes were made
- [x] Code diff has been done and been reviewed (it does not contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass
- [x] Code coverage has not gone down and all code touched or added is covered.
- [x] Code passes lint and prettier (hint: use `npm run test:plus` to run tests, lint, and prettier)
- [x] All dependent libraries are appropriately updated or have a corresponding PR related to this change
- [x] `cql4browsers.js` built with `npm run build:browserify` if source changed.

**Reviewer:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

File: src/elm/query.ts [hunk 1]
```diff
@@ -237,8 +237,33 @@ export class Query extends Expression {
   }
 
   exec(ctx: Context) {
+    let sourceResults = this.sources.map(s => s.expression.execute(ctx));
+    // If every source is null, the result is null
+    // See: https://jira.hl7.org/browse/FHIR-40225
+    if (sourceResults.every(r => r == null)) {
+      return null;
+    }
+    const isList = sourceResults.some(r => Array.isArray(r));
```
Comment (mgramigna) [inline src/elm/query.ts:246]: Should we re-use the `typeIsArray` util instead?

File: src/elm/query.ts [hunk 2]
```diff
@@ -237,8 +237,33 @@ export class Query extends Expression {
...
+    // For ease of processing, convert all the sources to lists and
+    // convert nulls to empty lists (we already handled the ALL null case)
+    sourceResults = sourceResults.map(r => {
+      if (r == null) {
+        return [];
+      } else if (Array.isArray(r)) {
```
Comment (mgramigna) [inline src/elm/query.ts:252]: Same as above. Not sure if this gives us any added benefit or not except consistency.

**Review by mgramigna — COMMENTED** [review]
Can confirm that this fixes the issue in fqm-execution (and some bonus cases as well after clarifying the spec further!)

I just have one small question, but LGTM overall

File: src/elm/query.ts (hunk 1 above)
Comment (cmoesel) [reply src/elm/query.ts:246]: These days, I think `Array.isArray` is pretty well-supported, so we should probably consider moving all uses of `typeIsArray` to `Array.isArray`.  Starting with this one. ;-)

File: src/elm/query.ts (hunk 2 above)
Comment (cmoesel) [reply src/elm/query.ts:252]: Nah.

**Review by mgramigna — APPROVED** [review]
Thanks! Glad we were able to clarify this

**Review by birick1 — APPROVED** [review]
Ugh ... O[n^2] operations for cartesian products.  No better suggestion.

**Pull request merged**